storage/merkle/tree.go

```
@@ -3,39 +3,103 @@
 // it makes insertions and deletions more complex, as we need to split nodes
-// and merge them, depending on whethere their are leaves or not.
+// and merge them, depending on whether there are leaves or not.
+//
+// CONVENTION:
+//  * If the tree contains _any_ elements, the tree is defined by its root vertex.
+//    This case follows completely the convention for nodes: "In any existing tree,
+//    all nodes are non-nil."
+//  * Without any stored elements, there exists no root vertex in this data model,
+//    and we set `root` to nil.
 type Tree struct {
-	root node
+	pathLength int
+	size       uint64
+	root       node
 }
 
 // NewTree creates a new empty patricia merkle tree.
 func NewTree() *Tree {
-	t := &Tree{}
-	return t
+	return &Tree{}
 }
 
-// Put will stores the given value in the trie under the given key. If the key
-// already exists, it will replace the value and return true.
-func (t *Tree) Put(key []byte, val interface{}) bool {
+// Size returns the number of stored key-value pairs
+func (t *Tree) Size() uint64 {
+	return t.size
+}
+
+// Put stores the given value in the trie under the given key. If the key
+// already exists, it will replace the value and return true. All inputs
+// are internally stored and copied where necessary, thereby allowing
+// external code to re-use the slices.
+// Returns:
+//  * (false, nil): key-value pair is stored; key did _not_ yet exist prior to update
+//  * (true, nil):  key-value pair is stored; key existed prior to update and the old
+//                  value was overwritten
+//  * (false, error): with possible error returns
+//    - ErrorIncompatibleKeyLength if `key` is empty, or exceeds the max supported length of 8192 bytes.
+//    - ErrorVariableKeyLengths if the provided key length is _different_ than previously stored elements.
+//    No other errors are returned.
+func (t *Tree) Put(key []byte, val []byte) (bool, error) {
+	if t.size == 0 { // empty key-value store
```

I like how the interface allows the same merkle tree to be re-used for keys of different lengths: once the tree is empty of all elements, the key length is set to the length of the first put element. This is a flexible interface though it comes with the cost of keeping track of `size`. 
I'm wondering if it's better to set the key length at the constructor level and not change it later. A new tree needs to be instantiated for a new key length. The advantage is to remove `size` and avoid updating it. 
I let you decide what is the better interface (maybe you have a use case in mind for the method `Size()` )

---

I like your suggestion of pre-defining the key length at construction time, as it reduces the complexity of the trie implementation.

---

Thanks for the refactor! Yes ignoring the error and adding the test is fine. This is similar to what you did for the Blake API 👌🏼